lib/elixir/lib/enum.ex

```diff
@@ -1900,32 +1900,53 @@ defmodule Enum do
   @spec slice(t, Range.t) :: list
   def slice(enumerable, range)
 
-  def slice(enumerable, first..last) when first >= 0 and last >= 0 do
-    # Simple case, which works on infinite enumerables
-    if last - first >= 0 do
-      slice(enumerable, first, last - first + 1)
-    else
-      []
+  def slice(enumerable, start..finish),
+    do: do_slice(enumerable, {start, finish})
+
+  defp do_slice(enumerable, {start, start}) do
+    case fetch(enumerable, start) do
+      :error ->
+        []
+      {:ok, result} ->
+        [result]
     end
   end
 
-  def slice(enumerable, first..last) do
-    {list, count} = enumerate_and_count(enumerable, 0)
-    corr_first = if first >= 0, do: first, else: first + count
-    corr_last = if last >= 0, do: last, else: last + count
-    length = corr_last - corr_first + 1
-    if corr_first >= 0 and length > 0 do
-      slice(list, corr_first, length)
+  defp do_slice(_enumerable, {start, finish})
+      when start > finish and ((start >= 0 and finish >= 0) or (start < 0 and finish < 0)) do
+    []
+  end
+
+  defp do_slice(enumerable, {start, finish}) when start < 0 or finish < 0 do
+    {enumerable, count} = get_enumerable_count(enumerable)
+
+    with  {:ok, normalized_start} when normalized_start < count <- normalize_index(start, count),
+          {:ok, normalized_finish} when normalized_start <= normalized_finish
+            <- normalize_index(finish, count) do
```

Comment:
I would rather have a longer line than break this line apart. Not sure what the rest of the core team thinks though?

It's also convention to put the binary operator (`<-`) on the preceding line.